### PR TITLE
fix(ingest/snowflake): Improve memory usage of metadata extraction

### PR DESCRIPTION
**Pull request by asikowitz**
Similar to https://github.com/datahub-project/datahub/pull/7315, attempts to reduce the memory usage of the snowflake connector by remove global data stores.

This differs from the bigquery changes, as I attempt to entirely remove the global stores, instead using the `lru_cache` decorator to cache the results of repeated function calls. We order our metadata extraction queries such that we should be doing no more queries while only caching the results of a single query at a time (vs. caching all the results). I have to keep a single store, `db_tables` to support per-database profiling, but make this per-database rather than fully global.

This does not include any lineage memory optimization -- I'll do that in a followup PR.

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/scripts/docgen.py [hunk 1]
```diff
@@ -679,8 +679,8 @@ def generate(
     i = 0
     for platform_id, platform_docs in sorted(
         source_documentation.items(),
-        key=lambda x: (x[1]['name'].casefold(), x[1]['name'])
-        if 'name' in x[1]
+        key=lambda x: (x[1]["name"].casefold(), x[1]["name"])
+        if "name" in x[1]
```
Comment (asikowitz) [inline metadata-ingestion/scripts/docgen.py:683]: Result of running `black`

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_profiler.py [hunk 2]
```diff
@@ -62,60 +59,54 @@ def get_workunits(
...
+            return
+
+        table_profile_requests = cast(List[TableProfilerRequest], profile_requests)
 
-                # We don't add to the profiler state if we only do table level profiling as it always happens
-                if self.state_handler:
-                    self.state_handler.add_to_state(
-                        dataset_urn, int(datetime.now().timestamp() * 1000)
-                    )
+        for request, profile in self.generate_profiles(
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_profiler.py:81]: No changes below, just changing indentation

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_report.py [hunk 3]
```diff
@@ -1,11 +1,13 @@
...
+from dataclasses import dataclass, field
+from typing import Dict, MutableSet, Optional
 
 from datahub.ingestion.source.snowflake.constants import SnowflakeEdition
 from datahub.ingestion.source.sql.sql_generic_profiler import ProfilingSqlReport
 from datahub.ingestion.source_report.sql.snowflake import SnowflakeReport
 from datahub.ingestion.source_report.usage.snowflake_usage import SnowflakeUsageReport
 
 
+@dataclass
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_report.py:10]: Pretty sure we want this on every class, even though it inherits from a dataclass?

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_report.py [hunk 4]
```diff
@@ -39,8 +47,8 @@ class SnowflakeV2Report(SnowflakeReport, SnowflakeUsageReport, ProfilingSqlRepor
 
     rows_zero_objects_modified: int = 0
 
-    _processed_tags: MutableSet[str] = set()
-    _scanned_tags: MutableSet[str] = set()
+    _processed_tags: MutableSet[str] = field(default_factory=set)
+    _scanned_tags: MutableSet[str] = field(default_factory=set)
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_report.py:51]: As a result of adding `@dataclass`

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py [hunk 5]
```diff
@@ -504,29 +489,36 @@ def get_workunits_internal(self) -> Iterable[MetadataWorkUnit]:
...
+            self.data_dictionary.get_columns_for_schema,
+            self.data_dictionary.get_pk_constraints_for_schema,
+            self.data_dictionary.get_fk_constraints_for_schema,
+        ]
+        for func in lru_cache_functions:
+            self.report.lru_cache_info[func.__name__] = func.cache_info()._asdict()
 
-        if self.config.profiling.enabled and len(databases) != 0:
-            yield from self.profiler.get_workunits(databases)
+        # TODO: The checkpoint state for stale entity detection can be committed here.
```
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py:509]: Just fixing a typo

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py [hunk 6]
```diff
@@ -504,29 +492,36 @@ def get_workunits_internal(self) -> Iterable[MetadataWorkUnit]:
                 yield from self._process_database(snowflake_db)
 
             except SnowflakePermissionError as e:
-                # FIXME - This may break satetful ingestion if new tables than previous run are emitted above
+                # FIXME - This may break stateful ingestion if new tables than previous run are emitted above
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py:495]: Thank you !

**Review by treff7es — APPROVED** [review]
nice, lgtm

**Review by mayurinehate — COMMENTED** [review]
Although this PR is merged, submitting comments as they are still valid. Probably, can be discussed offline and handled in followup PR, if it makes sense.

(I had added these comments earlier but they did not get published as I forgot to click "Submit Review" button at the end.)

File: metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py [hunk 7]
```diff
@@ -277,20 +277,8 @@ def __init__(self, ctx: PipelineContext, config: SnowflakeV2Config):
...
-        # TODO: In future, we may consider using queries and caching at database level first
-        self.schema_columns: Dict[
-            Tuple[str, str], Optional[Dict[str, List[SnowflakeColumn]]]
-        ] = {}
-        self.schema_pk_constraints: Dict[Tuple[str, str], Dict[str, SnowflakePK]] = {}
-        self.schema_fk_constraints: Dict[
-            Tuple[str, str], Dict[str, List[SnowflakeFK]]
-        ] = {}
+        # Caches tables for a single database. Consider moving to disk or S3 when possible.
+        self.db_tables: Dict[str, List[SnowflakeTable]] = {}
```
[the remark this answers is not in the document]
Comment (asikowitz) [reply metadata-ingestion/src/datahub/ingestion/source/snowflake/snowflake_v2.py:281]: Yup! I initially had neither, but had to add this one back for profiling

**Pull request merged**